auto_route/lib/src/router/controller/routing_controller.dart

```diff
@@ -657,17 +657,68 @@ class TabsRouter extends RoutingController {
   int? get previousIndex => _previousIndex;
 
   /// Updates [_activeIndex] and triggers a rebuild
-  void setActiveIndex(int index, {bool notify = true}) {
+  void setActiveIndex(int index, {bool notify = true}) async {
     assert(index >= 0 && index < _pages.length);
+
     if (_activeIndex != index) {
-      _previousIndex = _activeIndex;
-      _activeIndex = index;
-      if (notify) {
-        notifyAll();
+      final match = _routeMatch[index];
+      final result = await _canNavigate(match, onFailure: (_) {});
+      if (result.continueNavigation) {
+        _previousIndex = _activeIndex;
+        _activeIndex = index;
+        if (notify) {
+          notifyAll();
+        }
       }
     }
   }
 
+  late final AutoRouteGuard? _rootGuard =
+      (root is AutoRouteGuard) ? (root as AutoRouteGuard) : null;
+  Future<ResolverResult> _canNavigate(
+    RouteMatch route, {
+    OnNavigationFailure? onFailure,
+  }) async {
+    final guards = <AutoRouteGuard>[
+      if (_rootGuard != null) _rootGuard!,
+      ...route.guards,
+    ];
```

Comment:
The main porblem with reusing existing guards, is that it might cause side effects. I tried to fix this in my fork and faced the same challenged.

Consider the following:
- you have a page and this page has three tabs
- these tabs are using the same guard used on the root router
- within the root you have a a custom guard that auto redirects to another page.

with the code you're proposing, if you open the page with that same tab, your expected behavior is tfor auto_route to simply prevent navigating to that tab where its guard has `result.continueNavigation` set to false. Instead, your code will also trigger the redirect, breaking the flow.